pkg/dashboard/apiserver/event/event.go

```diff
@@ -97,6 +106,116 @@ func (s *Service) list(c *gin.Context) {
 	c.JSON(http.StatusOK, events)
 }
 
+// @Summary cascadeFetchEventsForWorkflow list all events for Workflow and related WorkflowNode.
+// @Description list all events for Workflow and related WorkflowNode.
+// @Tags events
+// @Produce json
+// @Param namespace query string false "The namespace of the object"
+// @Param object_id query string false "The UID of the Workflow"
+// @Param limit query string false "The max length of events list"
+// @Success 200 {array} core.Event
+// @Failure 500 {object} utils.APIError
+// @Router /events/workflow [get]
+func (s *Service) cascadeFetchEventsForWorkflow(c *gin.Context) {
+	ctx := c.Request.Context()
+	ns := c.Query("namespace")
+	objectId := c.Query("object_id")
+	limit := 0
+	limitString := c.Query("limit")
+	if len(limitString) > 0 {
+		parsedLimit, err := strconv.Atoi(limitString)
+		if err != nil {
+			u.SetAPIError(c, u.ErrBadRequest.Wrap(err, "parameter limit should be a integer"))
+			return
+		}
+		limit = parsedLimit
+	}
+
+	if ns == "" && !s.conf.ClusterScoped && s.conf.TargetNamespace != "" {
+		ns = s.conf.TargetNamespace
+
+		log.V(1).Info("Replace query namespace with", ns)
+	}
+
+	// we should fetch the events for Workflow and related WorkflowNode, so we need namespaced name at first
+	workflowEntity, err := s.workflowStore.FindByUID(ctx, objectId)
+	if err != nil {
+		if gorm.IsRecordNotFoundError(err) {
+			u.SetAPIError(c, u.ErrNotFound.Wrap(err, "this requested workflow is not found, object_id: %s", objectId))
+		} else {
+			u.SetAPIError(c, u.ErrInternalServer.WrapWithNoMessage(err))
+		}
+		return
+	}
+
+	// if workflow has been archived, the Workflow CR and WorkflowNode CR also has been deleted from kubernetes, it's
+	// no way to "cascade fetch" anymore
+	if workflowEntity.Archived {
+		u.SetAPIError(c, u.ErrBadRequest.New("this requested workflow already been archived, can not list events for it"))
+		return
+	}
+
+	kubeClient, err := clientpool.ExtractTokenAndGetClient(c.Request.Header)
+	if err != nil {
+		_ = c.Error(u.ErrBadRequest.WrapWithNoMessage(err))
```

Comment:
There are still some `c.Error`. k8s-related errors can be handled using `u.SetAPImachineryError(c, err)`.

Reply:
updated in 338a4e05, PTAL!